modules/transport-netty4/src/main/java/org/elasticsearch/transport/CopyBytesSocketChannel.java

```diff
@@ -0,0 +1,186 @@
+ */
+@SuppressForbidden(reason = "Channel#write")
+public class CopyBytesSocketChannel extends NioSocketChannel {
+
+    private static final int MAX_BYTES_PER_WRITE = 1 << 20;
+
+    private static final ThreadLocal<ByteBuffer> ioBuffer = ThreadLocal.withInitial(() -> ByteBuffer.allocateDirect(MAX_BYTES_PER_WRITE));
+    private final WriteConfig writeConfig = new WriteConfig();
+
+    public CopyBytesSocketChannel() {
+        super();
+    }
+
+    CopyBytesSocketChannel(Channel parent, SocketChannel socket) {
+        super(parent, socket);
+    }
+
+    @Override
+    protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+        SocketChannel ch = javaChannel();
+        int writeSpinCount = config().getWriteSpinCount();
+        do {
+            if (in.isEmpty()) {
+                // All written so clear OP_WRITE
+                clearOpWrite();
+                // Directly return here so incompleteWrite(...) is not called.
+                return;
+            }
+
+            // Ensure the pending writes are made of ByteBufs only.
+            int maxBytesPerGatheringWrite = writeConfig.getMaxBytesPerGatheringWrite();
+            ByteBuffer[] nioBuffers = in.nioBuffers(1024, maxBytesPerGatheringWrite);
+            assert Arrays.stream(nioBuffers).filter(Objects::nonNull).noneMatch(ByteBuffer::isDirect) : "Expected all to be heap buffers";
+            int nioBufferCnt = in.nioBufferCount();
+
+            if (nioBufferCnt == 0) {// We have something else beside ByteBuffers to write so fallback to normal writes.
+                writeSpinCount -= doWrite0(in);
+            } else {
+                // Zero length buffers are not added to nioBuffers by ChannelOutboundBuffer, so there is no need
+                // to check if the total size of all the buffers is non-zero.
+                ByteBuffer ioBuffer = getIoBuffer();
+                copyBytes(nioBuffers, nioBufferCnt, ioBuffer);
+                ioBuffer.flip();
+
+                int attemptedBytes = ioBuffer.remaining();
+                final int localWrittenBytes = ch.write(ioBuffer);
+                if (localWrittenBytes <= 0) {
+                    incompleteWrite(true);
+                    return;
+                }
+                adjustMaxBytesPerGatheringWrite(attemptedBytes, localWrittenBytes, maxBytesPerGatheringWrite);
+                in.removeBytes(localWrittenBytes);
+                --writeSpinCount;
+            }
+        } while (writeSpinCount > 0);
+
+        incompleteWrite(writeSpinCount < 0);
+    }
+
+    @Override
+    protected int doReadBytes(ByteBuf byteBuf) throws Exception {
+        final RecvByteBufAllocator.Handle allocHandle = unsafe().recvBufAllocHandle();
+        allocHandle.attemptedBytesRead(byteBuf.writableBytes());
+        ByteBuffer ioBuffer = getIoBuffer();
+        int bytesRead = javaChannel().read(ioBuffer);
+        ioBuffer.flip();
+        if (bytesRead > 0) {
+            byteBuf.writeBytes(ioBuffer);
```

Comment:
I'm thinking this line possibly expand `byteBuf` capacity. That behavior is different from `NioSocketChannel`.
Is this acceptable?

https://github.com/netty/netty/blob/netty-4.1.38.Final/transport/src/main/java/io/netty/channel/socket/nio/NioSocketChannel.java#L347

We experienced an issue after upgrading ES cluster from ES 7.1.1 to 7.4.2. When a bunch of index creation, like hundreds of indices, happens, master/data nodes start to leave from a cluster, and the cluster status becomes red. We are still investigating what is happening actually, and hopefully we can submit a report about the issue.

When we first ran ES 7.4.2, `jvm.options` does not include `-Dio.netty.allocator.numDirectArenas=0` line. After we observe an OOM issue, we added the flag. After that, the OOM issue does not happen. But, looks like the issue on index creation started.

Reply:
@occho since our `ioBuffer` is of fixed size and we are not using any read throttling at the moment, I believe this change did not introduce a new risk in terms of memory allocation. If anything, the change should make the memory use of the networking layer more less unpredictable. As you experienced yourself, setting `-Dio.netty.allocator.numDirectArenas=0` fixed any OOM issues. 

> We are still investigating what is happening actually, and hopefully we can submit a report about the issue.

Thanks for looking into that. If you feel like you don't have enough details to open a Github issue around a reproducible bug. Feel free to share what you have (logs etc.) on our [discuss forums](https://discuss.elastic.co/c/elasticsearch) for discussion.

Reply:
Understood. Thank you for clarifying that!